README.md
# Sync /etc/hosts to AWS Route 53

[![Build Status](https://travis-ci.org/claytono/sync-hosts-to-route53.svg?branch=master)](https://travis-ci.org/claytono/sync-hosts-to-route53) [![Go Report Card](https://goreportcard.com/badge/github.com/claytono/sync-hosts-to-route53)](https://goreportcard.com/report/github.com/claytono/sync-hosts-to-route53)

## What does this do?

This is a tool that will read a local file in `/etc/hosts` format and
synchronize the contents of that file to a specific domain in AWS Route 53.
The contents of the /etc/hosts file can be filtered such that only records in
specific networks will be synchronized.  By default the program will run
continuously, detecting changes in the hosts file automatically via inotify and
running periodic syncs to clean up any rogue Route 53 changes.

This was written to be used on Ubiquiti EdgeOS devices like the EdgeRouter
Lite.  This can be useful if you have public IP ranges hosted behind your
router, or if you have a VPN connection into the a private network.  This will
allow you to use any DNS caching resolver, instead of forcing DNS resolution
through a private DNSMasq instance.

There is nothing inherently specific to EdgeOS in this project, so if you need
something to sync an `/etc/hosts` file to Route53 once, or on a regular basis,
this might work well for you.

## What do I need to know?

* The local hosts file given is considered to be the source of truth for the
  networks  you specify to be managed.  Changes in Route 53 that don't match
  what is in the hosts file will be removed or overwritten.

* Host aliases in the input file are ignored.  Entries in `/etc/hosts` on
  EdgeOS devices that are added via DHCP never have alias entries.

* This app is written in Go, mostly because it generates binaries with no
  dependencies and it makes cross compiles for other architectures easy.

* Binaries can be found on the [GitHub releases
  page](https://github.com/claytono/sync-hosts-to-route53/releases)

* MIPS64 binaries are built and available on the [GitHub releases
  page](https://github.com/claytono/sync-hosts-to-route53/releases), but
  aren't recommended.  When tested on EdgeOS 1.9.1, inotify does not appear to
  work with the MIPS64 binary, but does work with the MIPS (32 bit) build.  In
  addition, when testing this in QEMU w/MIPS64, the binary just crashes on
  startup in the Golang GC code.

## AWS Authentication

This program uses the [official AWS golang
libary](https://github.com/aws/aws-sdk-go).  Because of this, any of the normal
methods of storing credentials for AWS that work with the CLI tools should work
for this program.  For example, you can set and export the `AWS_ACCESS_KEY_ID`
and `AWS_SECRET_ACCESS_KEY` environment variables.  More information can be
found in the "[SDK Configuration](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)"
section of the library documentation.

It is recommended that you set up an AWS IAM user specifically to be used with
this program.  That IAM user should be limited to only be able to add and
remove records in Route 53.  If you have multiple domains, you may wish to
limit that user to just the specific domain you wish to synchronize against. 

## Usage

Use `sync-hosts-to-route53 --help` for a complete list of available options.

The options available include:

### -m|--mode [oneshot|daemon]

This options must be either `oneshot` or `daemon`.  The default is `daemon`.

When run with `--mode daemon` or no `--mode` argument, the program will
synchronize the host file with Route 53 once, then setup inotify watches for
the host file, and synchronize automatically by default every 15 minutes.

When run with `--mode oneshot` the program will synchronize the host file
given with Route 53 once, then exit.

### -f|--file=HOSTFILE

This specifies the local hosts file to keep in sync with Route 53.  This should
be in the format of UNIX style `/etc/hosts` file.  This defaults to
`/etc/hosts`.

### --network=x.x.x.x/len

This option will direct the program to ignore all host entries in the local
file, or in the Route 53 domain that are not inside of the network blocks
given.  This option is required and can be specified more than once.  If you
wish to affect **all** entries in the domain, then you can specify `0.0.0.0/0`
to match all IP addresses.

### -d|--domain=

This specifies the Route 53 domain to synchronize with the local hosts file.
This option is required and has no default.

### -i|--interval=

How often to run the synchronization, even if no changes have been detected in
the local hosts file.  Local changes should be detected automatically, so this
mostly serves to correct any changes made in Route 53 that don't match the
local file.  This defaults to 15 minutes.  This is ignored in oneshot mode.

### --ttl=

This is the DNS record TTL in seconds to set on new Route 53 records.  This
defaults to 3600 seconds, or one hour.

### --no-qualify-hosts

By default the Route53 domain will be appended to the end of any host file
entries that appear to be lacking it.  To disable this behavior, specify
`--no-qualifiy-hosts`.

### --exclude-hosts

Exclude specific hosts from being synced to Route53.  This can be used to
prevent manually created items from being deleted during the sync process.
This can be specified multiple times.

### --no-wait

By default the program will wait for Route 53 updates to propagate after
submitting them.  To disable this behavior, specify `--no-wait`.

### --syslog

Enable logging to syslog in addition to stdout.

### --syslog-facility=

Syslog facility to log to.  Defaults to `user`.  See `syslog(3)` for list of
options.

### --syslog-only

Log just to syslog, skipping stdout entirely.  If this is specified then
`--syslog` is implied.  Note a single message will be logged to stdout
indicating all further output will go to syslog.

### --debug

Enable debug level logging.

### --version

Print version to stdout and exit.

### -h|--help

Print summary of these options

## Building From Source

Dependencies for this project are managed using [`go mod`](https://github.com/golang/go/wiki/Modules)

To build binaries for all platforms (used for releases), run `make
build-all-arch`.

## Author

Clayton O'Neill
clayton@oneill.net


CHANGELOG.md
# Changelog

## [1.1.5] - 2022-03-12
###
- Fixed bug with alias A records
- Switch from `glide` to `dep`

## [1.1.4] - 2019-05-05
###
- Rebuild binaries with 1.12.4
- Updated deps to latest stable

## [1.1.3] - 2017-10-23
###
- Rebuild binaries with 1.9.1

## [1.1.2] - 2017-08-02
### Changed
- Fixed issue with Route 53 errors causing daemon mode to exit.

## [1.1.1] - 2017-08-02
### Changed
- Fixed typo which made hosts not get synced.

## [1.1.0] - 2017-06-27
### Added
- New `--exclude-host` option to prevent deleting or changing Route 53 entries
  in allowed network ranges.

## [1.0.0] - 2017-06-15
- Initial version


host_test.go
package main

import (
	"net"
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

var hostCases = []struct {
	name     string
	unsorted hostList
	sorted   hostList
}{
	{name: "already sorted",
		unsorted: hostList{
			{hostname: "test1.com", ip: net.ParseIP("1.2.3.4")},
			{hostname: "test2.com", ip: net.ParseIP("1.2.3.4")}},
		sorted: hostList{
			{hostname: "test1.com", ip: net.ParseIP("1.2.3.4")},
			{hostname: "test2.com", ip: net.ParseIP("1.2.3.4")}},
	},
	{name: "reverse sorted",
		unsorted: hostList{
			{hostname: "test2.com", ip: net.ParseIP("1.2.3.4")},
			{hostname: "test1.com", ip: net.ParseIP("1.2.3.4")}},
		sorted: hostList{
			{hostname: "test1.com", ip: net.ParseIP("1.2.3.4")},
			{hostname: "test2.com", ip: net.ParseIP("1.2.3.4")}},
	},
	{name: "same name, different ip",
		unsorted: hostList{
			{hostname: "test.com", ip: net.ParseIP("1.2.3.5")},
			{hostname: "test.com", ip: net.ParseIP("1.2.3.4")}},
		sorted: hostList{
			{hostname: "test.com", ip: net.ParseIP("1.2.3.4")},
			{hostname: "test.com", ip: net.ParseIP("1.2.3.5")}},
	},
}

func TestSortHostList(t *testing.T) {
	for _, c := range hostCases {
		t.Run(c.name, func(t *testing.T) {
			sort.Sort(c.unsorted)
			assert.Equal(t, c.unsorted, c.sorted)
		})
	}
}


main_test.go
package main

import (
	"net"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestCompareHosts(t *testing.T) {
	cases := []struct {
		name     string
		hosts    hostList
		r53hosts hostList
		toUpdate hostList
		toDelete hostList
	}{
		{"noop",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{},
			hostList{},
		},
		{"need-update",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.5")},
			},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{},
		},
		{"add-new",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{},
		},
		{"remove-stale",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
				{hostname: "test2.test.com", ip: net.ParseIP("1.2.3.5")},
			},
			hostList{},
			hostList{
				{hostname: "test2.test.com", ip: net.ParseIP("1.2.3.5")},
			},
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			toUpdate, toDelete := compareHosts(c.hosts, c.r53hosts)
			assert.Equal(t, c.toUpdate, toUpdate)
			assert.Equal(t, c.toDelete, toDelete)
		})
	}
}

func TestRemoveExcludedHosts(t *testing.T) {
	cases := []struct {
		name     string
		hosts    hostList
		excludes []string
		result   hostList
	}{
		{"noop",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			[]string{"test2.test.com"},
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
		},
		{"remove-only-one",
			hostList{
				{hostname: "test1.test.com", ip: net.ParseIP("1.2.3.4")},
			},
			[]string{"test1.test.com"},
			hostList{},
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			result := removeExcludedHosts(c.hosts, c.excludes)
			assert.Equal(t, result, c.result)
		})
	}
}


route53_test.go
package main

import (
	"net"
	"testing"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/route53"
	"github.com/stretchr/testify/assert"
)

func TestConvertR53RecordsToHosts(t *testing.T) {
	input := []*route53.ResourceRecordSet{
		{
			Name: aws.String("test1.test.com"),
			Type: aws.String("A"),
			TTL:  aws.Int64(300),
			ResourceRecords: []*route53.ResourceRecord{
				{Value: aws.String("1.2.3.4")},
			},
		},
		{
			Name: aws.String("test2.test.com"),
			Type: aws.String("CNAME"),
			TTL:  aws.Int64(300),
			ResourceRecords: []*route53.ResourceRecord{
				{Value: aws.String("test1.test.com")},
			},
		},
		{
			Name: aws.String("test3.test.com"),
			Type: aws.String("A"),
			TTL:  aws.Int64(300),
			ResourceRecords: []*route53.ResourceRecord{
				{Value: aws.String("1.2.3.4")},
				{Value: aws.String("1.2.3.5")},
			},
		},
		{
			Name: aws.String("test4.test.com"),
			Type: aws.String("A"),
			TTL:  aws.Int64(300),
			ResourceRecords: []*route53.ResourceRecord{
				{Value: aws.String("abc")},
			},
		},
	}

	expected := hostList{{
		hostname: "test1.test.com",
		ip:       net.ParseIP("1.2.3.4"),
		rrset:    input[0],
	},
	}

	output := convertR53RecordsToHosts(input)
	assert.Equal(t, expected, output)

}


host.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"net"
	"os"
	"strings"

	"github.com/aws/aws-sdk-go/service/route53"
)

type hostEntry struct {
	hostname string
	ip       net.IP
	// Aliases are read from the /etc/hosts file, but not mapped to Route53
	aliases []string
	// rrset only exists for imported Route 53 records
	rrset *route53.ResourceRecordSet
}

type hostList []hostEntry

func (h hostList) Len() int {
	return len(h)
}

func (h hostList) Less(i, j int) bool {
	if h[i].hostname != h[j].hostname {
		return h[i].hostname < h[j].hostname
	}
	return bytes.Compare(h[i].ip, h[j].ip) < 0
}

func (h hostList) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func parseLine(line string) (*hostEntry, error) {
	if i := strings.Index(line, "#"); i >= 0 {
		line = line[0:i]
	}

	parts := strings.Fields(line)
	if len(parts) == 0 {
		return nil, nil
	}

	if len(parts) < 2 {
		return nil, fmt.Errorf("should contain at least two fields")
	}

	if ip := net.ParseIP(parts[0]); ip != nil {
		return &hostEntry{
			hostname: parts[1],
			ip:       ip,
			aliases:  parts[2:],
		}, nil
	}

	return nil, fmt.Errorf("%s is not a valid IP", parts[0])
}

func readHosts(filename string) (hosts hostList) {
	file, err := os.Open(filename)

	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	i := 0
	for scanner.Scan() {
		i++
		host, err := parseLine(scanner.Text())
		if err != nil {
			log.Warnf("%v on line %v, skipping\n", err, i)
			continue
		}
		if host != nil {
			host.hostname = canonifyHostname(host.hostname)
			hosts = append(hosts, *host)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return
}

func filterHostsByNetwork(hosts hostList, networks []CIDRNet) hostList {
	output := hostList{}
	for _, host := range hosts {
		for _, net := range networks {
			if net.Contains(host.ip) {
				output = append(output, host)
				break
			}
		}
	}
	return output
}

func qualifyHosts(hosts hostList, domain string) hostList {
	result := make(hostList, len(hosts))
	for i, h := range hosts {
		result[i] = h
		if !strings.HasSuffix(h.hostname, domain) {
			result[i].hostname = h.hostname + "." + domain
		}
	}

	return result
}


daemon.go
package main

import (
	"os"
	"path/filepath"
	"time"

	"github.com/Sirupsen/logrus"
	"github.com/rjeczalik/notify"
)

func setupNotify(filename string) (chan notify.EventInfo, string) {
	absfn, err := filepath.Abs(filename)
	if err != nil {
		log.Fatalf("Cannot convert %v to absolute path: %v", filename, err)
	}

	dir := filepath.Dir(absfn)
	basename := filepath.Base(absfn)

	cn := make(chan notify.EventInfo, 1)
	err = notify.Watch(dir, cn, notify.All)
	if err != nil {
		log.Fatal("Cannot setup watch for ", dir, ": ", err)
	}

	log.Infof("Watching %v for changes to %v", dir, basename)

	return cn, absfn
}

func runIfInputExists(filename string) {
	if _, err := os.Stat(filename); err != nil {
		log.Error("Cannot stat hosts file, skipping sync: ", err)
	} else {
		// Ignore errors here, since we want to keep retrying over and over.
		runOnce()
	}
}

func daemon(interval time.Duration, filename string) {
	cn, absfn := setupNotify(filename)
	defer notify.Stop(cn)

	log.Info("Running initial sync")
	runIfInputExists(absfn)

	log.Info("sync scheduled every ", interval)
	ticker := time.NewTicker(interval)

	for {
		resyncNeeded := false
		// Block on either the ticker or inotify
		select {
		case <-ticker.C:
			resyncNeeded = true
		case ei := <-cn:
			if ei.Path() == absfn {
				log.Info("file change event detected: ", ei)
				resyncNeeded = true
			} else if log.Level <= logrus.DebugLevel {
				// logrus levels are numerically the opposite of what you'd
				// expect.  This logs if debug logging is enabled, but not
				// otherwise.
				log.Debug("file change event detected: ", ei)
			}
		}

		if resyncNeeded {
			runIfInputExists(absfn)
		}
	}
}


route53.go
package main

import (
	"fmt"
	"net"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/route53"
	"github.com/pkg/errors"
)

type route53Client struct {
	sess *session.Session
	svc  *route53.Route53
}

func newRoute53() route53Client {
	var r53 route53Client

	// awsSession is global so that we only read the config once and reuse it
	r53.sess = session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))

	r53.svc = route53.New(r53.sess)

	return r53
}

func (r53 route53Client) getZone(domain string) (*route53.HostedZone, error) {
	params := &route53.ListHostedZonesByNameInput{
		DNSName: aws.String(domain),
	}
	resp, err := r53.svc.ListHostedZonesByName(params)
	if err != nil {
		return nil, errors.Wrap(err, "Cannot list zones")
	}

	if len(resp.HostedZones) == 0 {
		return nil, fmt.Errorf("could not find domain '%v'", domain)
	}

	if *resp.HostedZones[0].Name != (domain + ".") {
		return nil, fmt.Errorf("could not find domain '%v'", domain)
	}

	return resp.HostedZones[0], nil
}

func (r53 route53Client) getRecords(zid string) ([]*route53.ResourceRecordSet, error) {
	params := &route53.ListResourceRecordSetsInput{
		HostedZoneId: &zid,
	}
	resp, err := r53.svc.ListResourceRecordSets(params)
	if err != nil {
		return nil, errors.Wrap(err, "Cannot get records")
	}

	return resp.ResourceRecordSets, nil
}

func (r53 route53Client) getHosts(domain string) (hostList, error) {
	zone, err := r53.getZone(domain)
	if err != nil {
		return hostList{}, errors.Wrap(err, "Cannot get zone")
	}

	rawHosts, err := r53.getRecords(*zone.Id)
	if err != nil {
		return hostList{}, errors.Wrap(err, "Cannot get hosts")
	}

	return convertR53RecordsToHosts(rawHosts), nil
}

func (r53 route53Client) sync(domain string, ttl int64, wait bool, toUpdate []hostEntry, toDelete []hostEntry) error {
	zone, err := r53.getZone(domain)
	if err != nil {
		return errors.Wrap(err, "Cannot get zone")
	}

	changes := make([]*route53.Change, 0, len(toUpdate)+len(toDelete))
	for _, h := range toUpdate {
		change := route53.Change{
			Action: aws.String("UPSERT"),
			ResourceRecordSet: &route53.ResourceRecordSet{
				Name: aws.String(h.hostname),
				Type: aws.String("A"),
				TTL:  aws.Int64(ttl),
				ResourceRecords: []*route53.ResourceRecord{
					{Value: aws.String(h.ip.String())},
				},
			},
		}
		changes = append(changes, &change)
	}

	for _, h := range toDelete {
		change := route53.Change{
			Action:            aws.String("DELETE"),
			ResourceRecordSet: h.rrset,
		}
		changes = append(changes, &change)
	}

	input := route53.ChangeResourceRecordSetsInput{
		HostedZoneId: zone.Id,
		ChangeBatch: &route53.ChangeBatch{
			Changes: changes,
		},
	}

	log.Debug("Changeset for Route 53:")
	log.Debug(input)
	log.Infof("Adding/updating %v records, deleting %v out of date records",
		len(toUpdate), len(toDelete))

	resp, err := r53.svc.ChangeResourceRecordSets(&input)
	if err != nil {
		return errors.Wrapf(err, "Could not update Route 53 records")
	}

	if wait {
		log.Info("Waiting for Route 53 update to complete")
		gci := route53.GetChangeInput{
			Id: resp.ChangeInfo.Id,
		}
		err = r53.svc.WaitUntilResourceRecordSetsChanged(&gci)
		if err != nil {
			return errors.Wrapf(err, "Update failed during wait")
		}
		log.Info("Sync completed successfully")
	} else {
		log.Info("Sync queued for update")
	}

	return nil
}

func convertR53RecordsToHosts(rawHosts []*route53.ResourceRecordSet) hostList {
	hosts := hostList{}
	for _, rh := range rawHosts {
		if *rh.Type != "A" {
			continue
		}

		if len(rh.ResourceRecords) > 1 {
			log.Warnf("%v has too many resource records (%d), ignoring record",
				*rh.Name, len(rh.ResourceRecords))
			continue
		}
		if len(rh.ResourceRecords) == 0 {
			log.Debugf("%v has no resource records, ignoring record", *rh.Name)
			continue
		}
		ip := net.ParseIP(*rh.ResourceRecords[0].Value)
		if ip == nil {
			log.Warnf("cannot parse IP %v for %v, ignoring record",
				*rh.ResourceRecords[0].Value, *rh.Name)
			continue
		}
		host := hostEntry{
			hostname: canonifyHostname(*rh.Name),
			ip:       ip,
			rrset:    rh,
		}

		hosts = append(hosts, host)
	}

	return hosts
}


main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log/syslog"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/Sirupsen/logrus"
	logrus_syslog "github.com/Sirupsen/logrus/hooks/syslog"
	flags "github.com/jessevdk/go-flags"
	"github.com/pkg/errors"
)

// This will automatically be overridden at build time if the provided
// Makefile is used.
var version = "unknown"

var log = logrus.New()

var opts struct {
	Mode           string        `short:"m" long:"mode" description:"Operating mode" default:"daemon" choice:"daemon" choice:"oneshot"`
	File           string        `short:"f" long:"file" description:"Input file in /etc/hosts format" default:"/etc/hosts" value-name:"HOSTFILE"`
	Networks       []CIDRNet     `long:"network" description:"Filter by CIDR network" value-name:"x.x.x.x/len"`
	Domain         string        `short:"d" long:"domain" description:"Domain to update records in"`
	Interval       time.Duration `short:"i" long:"interval" description:"Seconds between scheduled resync times." default:"15m"`
	TTL            int64         `long:"ttl" description:"TTL to use for Route 53 records" default:"3600"`
	NoQualifyHosts bool          `long:"no-qualify-hosts" description:"Don't force domain to be added to end of hosts"`
	ExcludeHosts   []string      `long:"exclude-host" description:"Exclude one or more hosts from being synced"`
	NoWait         bool          `long:"no-wait" description:"Don't wait for Route 53 to finish update"`
	Syslog         bool          `long:"syslog" description:"Send logging to syslog in addition to stdout"`
	SyslogFacility string        `long:"syslog-facility" description:"Syslog facility to log under" default:"user"`
	SyslogOnly     bool          `long:"syslog-only" description:"Send logging *only* to syslog"`
	Debug          bool          `long:"debug" description:"Enable debug logging"`
	Version        bool          `long:"version" description:"Print version number and exit"`
}

func facilityStringToInt(facility string) syslog.Priority {
	lookup := map[string]syslog.Priority{
		"kern":     syslog.LOG_KERN,
		"user":     syslog.LOG_USER,
		"mail":     syslog.LOG_MAIL,
		"daemon":   syslog.LOG_DAEMON,
		"auth":     syslog.LOG_AUTH,
		"syslog":   syslog.LOG_SYSLOG,
		"lpr":      syslog.LOG_LPR,
		"news":     syslog.LOG_NEWS,
		"uucp":     syslog.LOG_UUCP,
		"cron":     syslog.LOG_CRON,
		"authpriv": syslog.LOG_AUTHPRIV,
		"ftp":      syslog.LOG_FTP,
		"local0":   syslog.LOG_LOCAL0,
		"local1":   syslog.LOG_LOCAL1,
		"local2":   syslog.LOG_LOCAL2,
		"local3":   syslog.LOG_LOCAL3,
		"local4":   syslog.LOG_LOCAL4,
		"local5":   syslog.LOG_LOCAL5,
		"local6":   syslog.LOG_LOCAL6,
		"local7":   syslog.LOG_LOCAL7,
	}

	p, ok := lookup[facility]

	if !ok {
		keys := make([]string, len(lookup))
		for k := range lookup {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		log.Errorf("\"%v\" is not a valid syslog facility", facility)
		log.Fatalln("Valid facilities are:", strings.Join(keys, " "))
	}

	return p
}

func parseOpts() {
	parser := flags.NewParser(&opts, flags.Default)

	if _, err := parser.Parse(); err != nil {
		if flagsErr, ok := err.(*flags.Error); ok && flagsErr.Type == flags.ErrHelp {
			os.Exit(0)
		} else {
			os.Exit(1)
		}
	}

	if opts.Version {
		fmt.Println("version: ", version)
		os.Exit(0)
	}

	if opts.Domain == "" {
		fmt.Fprintln(os.Stderr, "domain name must be specified (-d or --domain)")
		os.Exit(1)
	}

	if len(opts.Networks) == 0 {
		fmt.Fprintln(os.Stderr, "one or more networks must be provided (--network)")
		os.Exit(1)
	}

	// Accept trailing dot, but ignore it for consistency sake
	if strings.HasSuffix(opts.Domain, ".") {
		opts.Domain = opts.Domain[:len(opts.Domain)-1]
	}
}

func configureLogging() {
	if opts.SyslogOnly {
		log.Out = ioutil.Discard
	} else {
		// logrus defaults to stderr, but stdout is more conventional
		log.Out = os.Stdout
	}

	if opts.Debug {
		log.Level = logrus.DebugLevel
	}

	if opts.Syslog || opts.SyslogOnly {
		facility := facilityStringToInt(opts.SyslogFacility)
		log.Info("Disabling color for syslog")
		tf := &logrus.TextFormatter{DisableColors: true}
		log.Formatter = tf

		hook, err := logrus_syslog.NewSyslogHook("", "", facility, "sync-hosts-to-route53")
		if err == nil {
			log.Hooks.Add(hook)
		} else {
			log.Fatal("Cannot initialize syslog: ", err)
		}
	}

}

func canonifyHostname(hostname string) string {
	hostname = strings.ToLower(hostname)
	// Accept trailing dot, but ignore it for consistency sake
	if strings.HasSuffix(hostname, ".") {
		hostname = hostname[:len(hostname)-1]
	}

	return hostname
}

// compareHosts takes the contents of the local /etc/hosts file and the
// Route 53 hosts that should be compared and produces two arrays.  The first
// array is a list of Route 53 records that need to be updated, and the second
// array is a list of Route 53 records that should be deleted.
func compareHosts(hosts hostList, r53hosts hostList) (hostList, hostList) {
	// Build index on name, we'll delete entries out of here a we match them
	// against /etc/hosts entries.  The remaining entries aren't present
	// locally anymore and will need to be deleted.
	rhByName := map[string]hostEntry{}
	for _, rh := range r53hosts {
		rhByName[rh.hostname] = rh
	}

	toUpdate := hostList{}
	// Find existing hosts
	for _, h := range hosts {
		rh, ok := rhByName[h.hostname]
		if ok {
			delete(rhByName, h.hostname)
			if !h.ip.Equal(rh.ip) {
				toUpdate = append(toUpdate, h)
			}
		} else {
			toUpdate = append(toUpdate, h)
		}
	}

	toDelete := make(hostList, 0, len(rhByName))
	for _, rh := range rhByName {
		toDelete = append(toDelete, rh)
	}

	return toUpdate, toDelete
}

func removeDupes(hosts hostList) hostList {
	found := make(map[string]bool, len(hosts))

	// Sort hostlist to ensure stable duplication suppression.  We don't want
	// to ping pont between choosing different options because of parse order.
	sort.Sort(hosts)

	dupCount := 0
	result := make(hostList, 0, len(hosts))
	for _, h := range hosts {
		if _, ok := found[h.hostname]; ok {
			log.Warnf("Duplicate hostname found in hosts, ignoring (%v/%v)",
				h.hostname, h.ip.String())
			dupCount++
		} else {
			found[h.hostname] = true
			result = append(result, h)
		}
	}

	return result
}

func removeExcludedHosts(hosts hostList, excludeList []string) hostList {
	hl := make(hostList, 0, len(hosts))

	for _, h := range hosts {
		excluded := false
		for _, eh := range excludeList {
			if h.hostname == eh {
				excluded = true
			}
		}

		if !excluded {
			hl = append(hl, h)
		}
	}

	return hl
}

func runOnce() error {
	hosts := readHosts(opts.File)
	hosts = filterHostsByNetwork(hosts, opts.Networks)
	if !opts.NoQualifyHosts {
		hosts = qualifyHosts(hosts, opts.Domain)
	}
	hosts = removeDupes(hosts)
	hosts = removeExcludedHosts(hosts, opts.ExcludeHosts)

	r53 := newRoute53()
	r53Hosts, err := r53.getHosts(opts.Domain)
	if err != nil {
		log.Warn(errors.Wrap(err, "error when retrieving zones"))
		return err
	}
	r53Hosts = filterHostsByNetwork(r53Hosts, opts.Networks)
	r53Hosts = removeExcludedHosts(r53Hosts, opts.ExcludeHosts)

	toUpdate, toDelete := compareHosts(hosts, r53Hosts)
	if len(toUpdate) > 0 || len(toDelete) > 0 {
		if err := r53.sync(opts.Domain, opts.TTL, !opts.NoWait, toUpdate, toDelete); err != nil {
			log.Warn(errors.Wrap(err, "Could not sync records to Route 53"))
			return err
		}
	} else {
		log.Info("No changes needed.  Everything in sync.")
	}

	return nil
}

func main() {
	parseOpts()
	configureLogging()
	if opts.Mode == "oneshot" {
		runOnce()
	} else {
		daemon(opts.Interval, opts.File)
	}
}
